Flaky Test: TestMaxPendingChunkMessages

Example failure: https://github.com/apache/pulsar-client-go/actions/runs/4352893665/jobs/7606246572

Log:
```
2023-03-07T10:15:51.9874801Z === RUN   TestMaxPendingChunkMessages
2023-03-07T10:15:51.9875376Z time="2023-03-07T10:12:36Z" level=info msg="Connecting to broker" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9875933Z time="2023-03-07T10:12:36Z" level=info msg="TCP connection established" local_addr="127.0.0.1:40150" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9876586Z time="2023-03-07T10:12:36Z" level=info msg="Connection is ready" local_addr="127.0.0.1:40150" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9877268Z time="2023-03-07T10:12:36Z" level=info msg="Connected producer" cnx="127.0.0.1:40150 -> 127.0.0.1:6650" epoch=0 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9878089Z time="2023-03-07T10:12:36Z" level=info msg="Created producer" cnx="127.0.0.1:40150 -> 127.0.0.1:6650" producerID=1 producer_name=standalone-0-422 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9878584Z time="2023-03-07T10:12:36Z" level=info msg="Connecting to broker" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9879427Z time="2023-03-07T10:12:36Z" level=info msg="TCP connection established" local_addr="127.0.0.1:40160" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9879941Z time="2023-03-07T10:12:36Z" level=info msg="Connection is ready" local_addr="127.0.0.1:40160" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9880694Z time="2023-03-07T10:12:36Z" level=info msg="Connected producer" cnx="127.0.0.1:40160 -> 127.0.0.1:6650" epoch=0 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9881600Z time="2023-03-07T10:12:36Z" level=info msg="Created producer" cnx="127.0.0.1:40160 -> 127.0.0.1:6650" producerID=1 producer_name=standalone-0-424 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9882066Z time="2023-03-07T10:12:36Z" level=info msg="Connecting to broker" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9882598Z time="2023-03-07T10:12:36Z" level=info msg="TCP connection established" local_addr="127.0.0.1:40164" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9883107Z time="2023-03-07T10:12:36Z" level=info msg="Connection is ready" local_addr="127.0.0.1:40164" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9883696Z time="2023-03-07T10:12:36Z" level=info msg="Connected producer" cnx="127.0.0.1:40164 -> 127.0.0.1:6650" epoch=0 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9884377Z time="2023-03-07T10:12:36Z" level=info msg="Created producer" cnx="127.0.0.1:40164 -> 127.0.0.1:6650" producerID=1 producer_name=standalone-0-425 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9885174Z time="2023-03-07T10:12:36Z" level=info msg="Connecting to broker" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9885723Z time="2023-03-07T10:12:36Z" level=info msg="TCP connection established" local_addr="127.0.0.1:40170" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9886251Z time="2023-03-07T10:12:36Z" level=info msg="Connection is ready" local_addr="127.0.0.1:40170" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9886855Z time="2023-03-07T10:12:36Z" level=info msg="Connected producer" cnx="127.0.0.1:40170 -> 127.0.0.1:6650" epoch=0 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9887558Z time="2023-03-07T10:12:36Z" level=info msg="Created producer" cnx="127.0.0.1:40170 -> 127.0.0.1:6650" producerID=1 producer_name=standalone-0-426 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9888033Z time="2023-03-07T10:12:36Z" level=info msg="Connecting to broker" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9888698Z time="2023-03-07T10:12:36Z" level=info msg="TCP connection established" local_addr="127.0.0.1:40182" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9889313Z time="2023-03-07T10:12:36Z" level=info msg="Connection is ready" local_addr="127.0.0.1:40182" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9889864Z time="2023-03-07T10:12:36Z" level=info msg="Connected producer" cnx="127.0.0.1:40182 -> 127.0.0.1:6650" epoch=0 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9890492Z time="2023-03-07T10:12:36Z" level=info msg="Created producer" cnx="127.0.0.1:40182 -> 127.0.0.1:6650" producerID=1 producer_name=standalone-0-427 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9891034Z time="2023-03-07T10:12:36Z" level=info msg="Connecting to broker" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9891576Z time="2023-03-07T10:12:36Z" level=info msg="TCP connection established" local_addr="127.0.0.1:40192" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9892248Z time="2023-03-07T10:12:36Z" level=info msg="Connection is ready" local_addr="127.0.0.1:40192" remote_addr="pulsar://localhost:6650"
2023-03-07T10:15:51.9893004Z time="2023-03-07T10:12:36Z" level=info msg="Connected consumer" consumerID=1 name=spsfq subscription=chunk-subscriber topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9893763Z time="2023-03-07T10:12:36Z" level=info msg="Created consumer" consumerID=1 name=spsfq subscription=chunk-subscriber topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9894060Z     message_chunking_test.go:236: 
2023-03-07T10:15:51.9895061Z         	Error Trace:	/pulsar/pulsar-client-go/pulsar/message_chunking_test.go:236
2023-03-07T10:15:51.9895399Z         	Error:      	Should not be: 200
2023-03-07T10:15:51.9895709Z         	Test:       	TestMaxPendingChunkMessages
2023-03-07T10:15:51.9896377Z time="2023-03-07T10:12:37Z" level=info msg="Closing consumer=1" consumerID=1 name=spsfq subscription=chunk-subscriber topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9897045Z time="2023-03-07T10:12:37Z" level=info msg="Closed consumer" consumerID=1 name=spsfq subscription=chunk-subscriber topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9897748Z time="2023-03-07T10:12:37Z" level=info msg="close consumer, exit reconnect" consumerID=1 name=spsfq subscription=chunk-subscriber topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9898414Z time="2023-03-07T10:12:37Z" level=info msg="Closing producer" producerID=1 producer_name=standalone-0-422 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9899381Z time="2023-03-07T10:12:37Z" level=info msg="Closed producer" producerID=1 producer_name=standalone-0-422 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9900028Z time="2023-03-07T10:12:37Z" level=info msg="close producer, exit reconnect" producerID=1 producer_name=standalone-0-422 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9900700Z time="2023-03-07T10:12:37Z" level=info msg="Closing producer" producerID=1 producer_name=standalone-0-424 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9901256Z time="2023-03-07T10:12:37Z" level=info msg="Closed producer" producerID=1 producer_name=standalone-0-424 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9901844Z time="2023-03-07T10:12:37Z" level=info msg="close producer, exit reconnect" producerID=1 producer_name=standalone-0-424 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9902408Z time="2023-03-07T10:12:37Z" level=info msg="Closing producer" producerID=1 producer_name=standalone-0-425 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9902973Z time="2023-03-07T10:12:37Z" level=info msg="Closed producer" producerID=1 producer_name=standalone-0-425 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9903564Z time="2023-03-07T10:12:37Z" level=info msg="close producer, exit reconnect" producerID=1 producer_name=standalone-0-425 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9904129Z time="2023-03-07T10:12:37Z" level=info msg="Closing producer" producerID=1 producer_name=standalone-0-426 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9904673Z time="2023-03-07T10:12:37Z" level=info msg="Closed producer" producerID=1 producer_name=standalone-0-426 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9905262Z time="2023-03-07T10:12:37Z" level=info msg="close producer, exit reconnect" producerID=1 producer_name=standalone-0-426 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9906087Z time="2023-03-07T10:12:37Z" level=info msg="Closing producer" producerID=1 producer_name=standalone-0-427 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9906768Z time="2023-03-07T10:12:37Z" level=info msg="Closed producer" producerID=1 producer_name=standalone-0-427 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9907404Z time="2023-03-07T10:12:37Z" level=info msg="close producer, exit reconnect" producerID=1 producer_name=standalone-0-427 topic="persistent://public/default/my-topic-38610081"
2023-03-07T10:15:51.9907733Z --- FAIL: TestMaxPendingChunkMessages (1.25s)
```

--- a/pulsar/message_chunking_test.go
+++ b/pulsar/message_chunking_test.go
@@ -151,8 +151,6 @@ func TestLargeMessage(t *testing.T) {
 }
 
 func TestMaxPendingChunkMessages(t *testing.T) {
-	rand.Seed(time.Now().Unix())
-
 	client, err := NewClient(ClientOptions{
 		URL: lookupURL,
 	})
@@ -562,13 +560,18 @@ func sendSingleChunk(p Producer, uuid string, chunkID int, totalChunks int) {
 	mm.ChunkId = proto.Int32(int32(chunkID))
 	producerImpl.updateMetadataSeqID(mm, msg)
 
+	doneCh := make(chan struct{})
 	producerImpl.internalSingleSend(
 		mm,
 		msg.Payload,
 		&sendRequest{
-			callback: func(id MessageID, producerMessage *ProducerMessage, err error) {},
-			msg:      msg,
+			callback: func(id MessageID, producerMessage *ProducerMessage, err error) {
+				close(doneCh)
+			},
+			msg: msg,
 		},
 		uint32(internal.MaxMessageSize),
 	)
+
+	<-doneCh
 }

--- a/pulsar/message_chunking_test.go
+++ b/pulsar/message_chunking_test.go
@@ -24,9 +24,12 @@ import (
 	"math/rand"
 	"net/http"
 	"strings"
-	"sync"
 	"testing"
 	"time"
+
+	"github.com/apache/pulsar-client-go/pulsar/internal"
+
+	"google.golang.org/protobuf/proto"
 
 	"github.com/stretchr/testify/assert"
 )
@@ -157,83 +160,52 @@ func TestMaxPendingChunkMessages(t *testing.T) {
 	defer client.Close()
 
 	topic := newTopicName()
+	producer, err := client.CreateProducer(ProducerOptions{
+		Topic:               topic,
+		DisableBatching:     true,
+		EnableChunking:      true,
+		ChunkMaxMessageSize: 10,
+	})
+	assert.NoError(t, err)
+	assert.NotNil(t, producer)
 
-	totalProducers := 5
-	producers := make([]Producer, 0, 20)
-	defer func() {
-		for _, p := range producers {
-			p.Close()
-		}
-	}()
-
-	clients := make([]Client, 0, 20)
-	defer func() {
-		for _, c := range clients {
-			c.Close()
-		}
-	}()
-
-	for j := 0; j < totalProducers; j++ {
-		pc, err := NewClient(ClientOptions{
-			URL: lookupURL,
-		})
-		assert.Nil(t, err)
-		clients = append(clients, pc)
-		producer, err := pc.CreateProducer(ProducerOptions{
-			Topic:               topic,
-			DisableBatching:     true,
-			EnableChunking:      true,
-			ChunkMaxMessageSize: 10,
-		})
-		assert.NoError(t, err)
-		assert.NotNil(t, producer)
-		producers = append(producers, producer)
-	}
-
-	consumer, err := client.Subscribe(ConsumerOptions{
+	c, err := client.Subscribe(ConsumerOptions{
 		Topic:                    topic,
 		Type:                     Exclusive,
 		SubscriptionName:         "chunk-subscriber",
 		MaxPendingChunkedMessage: 1,
 	})
 	assert.NoError(t, err)
-	assert.NotNil(t, consumer)
-	defer consumer.Close()
+	assert.NotNil(t, c)
+	defer c.Close()
+	pc := c.(*consumer).consumers[0]
 
-	totalMsgs := 40
-	wg := sync.WaitGroup{}
-	wg.Add(totalMsgs * totalProducers)
-	for i := 0; i < totalMsgs; i++ {
-		for j := 0; j < totalProducers; j++ {
-			p := producers[j]
-			go func() {
-				ID, err := p.Send(context.Background(), &ProducerMessage{
-					Payload: createTestMessagePayload(50),
-				})
-				assert.NoError(t, err)
-				assert.NotNil(t, ID)
-				wg.Done()
-			}()
-		}
-	}
-	wg.Wait()
+	sendSingleChunk(producer, "0", 0, 2)
+	// MaxPendingChunkedMessage is 1, the chunked message with uuid 0 will be discarded
+	sendSingleChunk(producer, "1", 0, 2)
 
-	received := 0
-	for i := 0; i < totalMsgs*totalProducers; i++ {
-		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
-		msg, err := consumer.Receive(ctx)
-		cancel()
-		if msg == nil || (err != nil && errors.Is(err, context.DeadlineExceeded)) {
-			break
-		}
+	// chunkedMsgCtx with uuid 0 should be discarded
+	retryAssert(t, 3, 200, func() {}, func(t assert.TestingT) bool {
+		pc.chunkedMsgCtxMap.mu.Lock()
+		defer pc.chunkedMsgCtxMap.mu.Unlock()
+		return assert.Equal(t, 1, len(pc.chunkedMsgCtxMap.chunkedMsgCtxs))
+	})
 
-		received++
+	sendSingleChunk(producer, "1", 1, 2)
 
-		err = consumer.Ack(msg)
-		assert.NoError(t, err)
-	}
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	msg, err := c.Receive(ctx)
+	cancel()
 
-	assert.NotEqual(t, totalMsgs*totalProducers, received)
+	assert.NoError(t, err)
+	assert.Equal(t, "chunk-1-0|chunk-1-1|", string(msg.Payload()))
+
+	// Ensure that the chunked message of uuid 0 is discarded.
+	sendSingleChunk(producer, "0", 1, 2)
+	ctx, cancel = context.WithTimeout(context.Background(), 5*time.Second)
+	msg, err = c.Receive(ctx)
+	cancel()
+	assert.True(t, errors.Is(err, context.DeadlineExceeded))
 }
 
 func TestExpireIncompleteChunks(t *testing.T) {
@@ -575,4 +547,28 @@ func createTestMessagePayload(size int) []byte {
 		payload[i] = byte(rand.Intn(100))
 	}
 	return payload
+}
+
+//nolint:all
+func sendSingleChunk(p Producer, uuid string, chunkID int, totalChunks int) {
+	msg := &ProducerMessage{
+		Payload: []byte(fmt.Sprintf("chunk-%s-%d|", uuid, chunkID)),
+	}
+	producerImpl := p.(*producer).producers[0].(*partitionProducer)
+	mm := producerImpl.genMetadata(msg, len(msg.Payload), time.Now())
+	mm.Uuid = proto.String(uuid)
+	mm.NumChunksFromMsg = proto.Int32(int32(totalChunks))
+	mm.TotalChunkMsgSize = proto.Int32(int32(len(msg.Payload)))
+	mm.ChunkId = proto.Int32(int32(chunkID))
+	producerImpl.updateMetadataSeqID(mm, msg)
+
+	producerImpl.internalSingleSend(
+		mm,
+		msg.Payload,
+		&sendRequest{
+			callback: func(id MessageID, producerMessage *ProducerMessage, err error) {},
+			msg:      msg,
+		},
+		uint32(internal.MaxMessageSize),
+	)
 }

